### DISCUSSION TITLE
On TestCases...

### DISCUSSION THREAD

**Discussion by JustinGrote** [pester/Pester#1997] [Q&A] [answered]
I noticed that testcases return multiple results in both NoRun and Run mode. 

Two questions:
1. Why can't the label of the test be updated to interpolate <Name> wildcards to the actual name of the value used?
2. Can the exception TargetObject be updated to point to the line where the testcase originated? This would make it easier to lens to which particular test case is not working.

Example:

```powershell
Describe "Various Test Styles" {
    # TODO: Handle this scenario. Today it currently just accepts the first result
    It "Multiple TestCases Array <Name>" {
        $Name | Should -BeOfType [int]
    } -TestCases @(
        @{Name = 1}
        @{Name = 'Should Fail'}
        @{Name = 2}
    )
}
```

Translated JSON Output by PesterInterface.ps1
```
[
  {
    "type": "test",
    "id": "C:\\Users\\JGrote\\Projects\\vscode-pester-test-adapter\\sample\\Tests\\TestStyles.Tests.ps1:3",
    "file": "C:\\Users\\JGrote\\Projects\\vscode-pester-test-adapter\\sample\\Tests\\TestStyles.Tests.ps1",
    "startLine": 2,
    "endLine": 4,
    "label": "Multiple TestCases Array <Name>",
    "result": 3,
    "duration": 8,
    "message": null,
    "expected": null,
    "actual": null,
    "targetFile": null,
    "targetLine": -1
  },
  {
    "type": "test",
    "id": "C:\\Users\\JGrote\\Projects\\vscode-pester-test-adapter\\sample\\Tests\\TestStyles.Tests.ps1:3",
    "file": "C:\\Users\\JGrote\\Projects\\vscode-pester-test-adapter\\sample\\Tests\\TestStyles.Tests.ps1",
    "startLine": 2,
    "endLine": 4,
    "label": "Multiple TestCases Array <Name>",
    "result": 4,
    "duration": 47,
    "message": "Expected the value to have type [int] or any of its subtypes, but got 'Should Fail' with type [string].",
    "expected": "the value to have type [int] or any of its subtypes",
    "actual": "'Should Fail' with type [string]",
    "targetFile": "C:\\Users\\JGrote\\Projects\\vscode-pester-test-adapter\\sample\\Tests\\TestStyles.Tests.ps1",
    "targetLine": 3
  },
  {
    "type": "test",
    "id": "C:\\Users\\JGrote\\Projects\\vscode-pester-test-adapter\\sample\\Tests\\TestStyles.Tests.ps1:3",
    "file": "C:\\Users\\JGrote\\Projects\\vscode-pester-test-adapter\\sample\\Tests\\TestStyles.Tests.ps1",
    "startLine": 2,
    "endLine": 4,
    "label": "Multiple TestCases Array <Name>",
    "result": 3,
    "duration": 2,
    "message": null,
    "expected": null,
    "actual": null,
    "targetFile": null,
    "targetLine": -1
  }
]
```

**Comment by fflaten** [comment, marked as answer]
> 1. Why can't the label of the test be updated to interpolate <Name> wildcards to the actual name of the value used?

The lazy variables may reference other variables available at runtime than those in testcase, so we'd have to run all BeforeAll/BeforeEach setups as well which can be time-consuming. SkipRun currently skips the entire Run-phase to avoid that.

> 2. Can the exception TargetObject be updated to point to the line where the testcase originated? 

Testcases can be an array that's reused in multiple tests, even with similar or the same name. Wouldn't that make it unclear which test you're referencing?

I do see you problem though as discussed earlier. It's hard to reference data-driven tests based due to script parameters and x levels of foreach/testcases. Best option is likely containerId:startline:testcaseindex (containerId being scritblock-id or path) and hope it's not so dynamic it changes on the next run. 😁

**Reply by JustinGrote** [reply]
For my purposes that's fine, it's very loose tracking based on a unique ID. However, based on the pester result object, for those individual test case objects, I have no idea how to distinguish between them. What object in the Pester class would work?

**Comment by JustinGrote** [comment]
It turns out that when doing discovery with NoRun, the Data property is populated! So it is possible to distinguish testcases. I made a function to generate unique test IDs even for TestCase Items. Lots of edge case issues but its good enough for now.

Function
```powershell
function New-TestItemId {
    <#
    .SYNOPSIS
    Create a string that uniquely identifies a test or test suite
    .NOTES
    Can be replaced with expandedpath if https://github.com/pester/Pester/issues/2005 is fixed
    #>
    [CmdletBinding()]
    param(
        [Parameter(Mandatory,ValueFromPipeline)]
        [ValidateScript({
            $null -ne ($_.PSTypeNames -match '^(Pester\.)?Test$')
        })]$Test
    )
    process {
        if ($Test.Path -match '\|') {
            throw [NotSupportedException]"The pipe character '|' is not supported in test names with this adapter. Please remove all pipes from test/context/describe names"
        }

        #TODO: Edge cases
        #Non-String Object
        #Non-STring Hashtable
        #Other dictionaries
        #Nested Hashtables
        #Fancy TestCases, maybe just iterate them as TestCaseN or exclude
        $Data = $Test.Data
        if ($Data) {
            if ($Data -is [IDictionary]) {
                $Test.Data.GetEnumerator() | Sort-Object Name | Foreach-Object {
                    $Test.Path += [String]([String]$PSItem.Name + '=' + [String]$PSItem.Value)
                }
            } else {
                $Test.Path += [string]$Data
            }
        }

        $Test.Path -join '|'
    }

}
```

Result on several tests including testcases.
```powershell
➜ (invoke-pester -Configuration @{Run=@{SkipRun=$true;PassThru=$true}}).tests | New-TestItemId
```
```
Test|test
Basic|Describe-Level Succeeds
Basic|Describe-Level Fails
Basic|Succeeds|True
Basic|Succeeds|False
Basic|Succeeds|ShouldBeTrue
Basic|Fails|Throws
Basic|Fails|ShouldThrow
Basic|Fails|ShouldBeTrue
Basic|Fails|ShouldBeGreaterThan
TestCases|TestCase Array <_>|1
TestCases|TestCase Array <_>|2
TestCases|TestCase Array <_>|red
TestCases|TestCase Array <_>|blue
TestCases|TestCase HashTable <Name>|Name=1
TestCases|TestCase HashTable <Name>|Name=2
TestCases|TestCase HashTable <Name>|Name=red
TestCases|TestCase HashTable <Name>|Name=blue
Duplicate DescribeWithContext|DupeContext|DupeContext
Duplicate DescribeWithContext|DupeContext|DupeContext
```

**Reply by JustinGrote** [reply]
Result is the same with both SkipRun and Regular Run, so I can correlate the test cases to their test runs. Doesn't address the targetObject issue but at least testcases can be enumerated in a test adapter now using this technique.

**Reply by fflaten** [reply]
Cool! Yeah Data is populated. 

Have you looked into -ForEach on context/describe yet? Will cause duplicate tests and testcases.

**Reply by nohwnd** [reply]
If you look into Pester source, there should be remnants of index based ids for test cases from the first version of the runtime. That will give you the same basic capabilities, with similar tradeoffs, but you won't have to expand the names twice (one partially on discovery, and once fully on runtime). It should also work reasonable well for cases where there is foreach on parent block. For generated tests there should be less zombie tests. You could also distinguish narrow runs (.e.g 1 test, 1 block) and wide runs (whole file), and act accordingly, to avoid: 1) dumping all existing test results for narrow run, and 2) keeping zombie results for wide runs.

**Reply by nohwnd** [reply]
And you should also be able to order tests based on their position in test cases. Even though that looks like it is working already pretty well.

**Reply by JustinGrote** [reply]
@nohwnd yeah part of my goal is that test history is tracked, and I wanted to preserve that as much as possible if a test "moves" which is why I wanted to go beyond simple file:line matching.

**Reply by JustinGrote** [reply]
@fflaten haven't tried it yet, right now I'm just dropping duplicates with an error message.

**Reply by JustinGrote** [reply]
@fflaten it doesn't populate the data but I noticed .block.data has the right info so I think I can make it work, pending...

**Reply by JustinGrote** [reply]
@fflaten got it to work, the unicode characters aren't coming over but that's a separate issue.
![image](https://user-images.githubusercontent.com/15258962/122589951-bc7aff00-d015-11eb-89a3-fa1ecacfbb38.png)

**Comment by JustinGrote** [comment]
Implemented in https://github.com/JustinGrote/vscode-pester-test-adapter/commit/3e6f6b94e27ac90dd0794d13f87bf0984f63e533

And I am able to correlate the tests together using my constructed id.
![image](https://user-images.githubusercontent.com/15258962/122516726-f9b6a100-cfc3-11eb-8b49-c8324cec5efc.png)
